Update OoT and MM's cutscene names

Yanis002 opened the pull request:
This also uses the enums for the types where it fits, currently it's hardcoded but I will use the config files later. Since this is the only thing left to do I was thinking why not opening the PR to get some feedback on things to improve. Also I tried to be accurate to both decomps to keep things understandable

Yanis002 committed:
progress

Yanis002 committed:
fixed non-matching issues

Yanis002 committed:
moved the enums to their own file & improvements

Yanis002 committed:
format

Yanis002 committed:
split MM and OoT

Yanis002 committed:
more cleanup

Yanis002 committed:
initial MM update

Yanis002 committed:
MM improvements

Yanis002 committed:
format

Yanis002 committed:
class/function name update

Yanis002 committed:
reduce the amount of zeroes (match)

Yanis002 committed:
enum rename

Yanis002 committed:
format

Yanis002 committed:
Merge remote-tracking branch 'upstream/master' into cs_update

Yanis002 committed:
removed hardcoded enums for OoT

Yanis002 commented:
I removed the hardcoded enums for OoT and instead I'm using a config file, it's an XML file since it makes things easier to manage imo, [here's](https://gist.github.com/Yanis42/b9c0480459b46d1b88d8efe1d341a881) how the file looks like, I'll port these changes to MM later

Edit: [here's](https://gist.github.com/Yanis42/bfc87af02429fb98343c99cdcdbce8f0) MM's XML file (and for both config files just add ``<CutsceneEnumData File="CutsceneEnumData.xml"/>``)

louist103 commented on ZAPD/OtherStructs/CutsceneMM_EnumData.h:
```diff
@@ -0,0 +1,200 @@
+
+// see
+// https://github.com/zeldaret/mm/blob/0c7b90cf97f26483c8b6a98ae099a295f61e72ab/include/z64cutscene.h#L109-118
+const std::string csMMModifySeqTypes[] = {
```
Unless I missed something these are only accessed with `.c_str()` so they can just be arrays of c strings.

Yanis002 replied:
this file has been deleted since (replaced by xml magic) though I get bad results when using ``char*`` inside the map from ``CutsceneEnumData``, the output is wrong and idk why so I just used c_str instead, let me know if there's a fix or not (I forgot a lot about C++ over the years lol)

louist103 commented on ZAPD/OtherStructs/CutsceneOoT_Commands.cpp:
```diff
@@ -0,0 +1,455 @@
+
+	if (commandID == (uint32_t)CutsceneOoT_CommandType::CS_CMD_CAM_AT_SPLINE)
+	{
+		listStr = "CS_CAM_AT_SPLINE";
```
There can also be c strings.

louist103 commented on ZAPD/GameConfig.cpp:
```diff
@@ -165,6 +165,54 @@ void GameConfig::ConfigFunc_ExternalFile(const tinyxml2::XMLElement& element)
+			std::string itemID = item->Attribute("ID");
+			uint16_t itemIndex = atoi(item->Attribute("Index"));
+
+			if (!strcmp(enumKey.c_str(), "cmd"))
```
Either make `enumKey` a c string or use the `==` operator.

louist103 commented on ZAPD/GameConfig.h:
```diff
@@ -21,6 +21,18 @@ class ExternalFile
 };
 
+// Stores data from the XML file, the integer is the index (via ATOI) and the string is the value
+class CutsceneData
```
This should be named something that doesn't make someone think its the actual data.

Yanis002 replied:
I renamed that to ``CutsceneEnumData``, let me know if you think of a better name

louist103 reviewed (commented):
Looks good so far. I will take a more in depth review when its ready. Here is a few things to get started with.

Yanis002 committed:
review 1

Yanis002 committed:
remove mm hardcoded enums

Yanis002 committed:
format

Yanis002 commented:
I think this is ready for review now, I did everything I had in mind for cutscenes I believe (also sorry for the big diff)

Yanis002 committed:
ActorCutscene -> CutsceneEntry
CutsceneEntry -> CutsceneScriptEntry

Yanis002 committed:
format

Yanis002 committed:
some changes in (actor) cutsceneentry

Yanis002 commented:
new changes since last time: 
- renamed ActorCutscene and old CutsceneEntry
- use enum names (from ``CutsceneEndSfx``) for end sfx inside CutsceneEntry
- renamed the variables to match new CutsceneEntry
- use spawn flag macro when possible
- format the files and renamed "SetActorCutsceneList.h/.cpp" to "SetCutsceneEntryList.h/.cpp"

I didn't change the scene command since it's not changed yet in decomp (I was hesitating, also the config gist has been updated), I don't think I'm gonna do more changes now (for real this time lol)

Yanis002 committed:
config "CutsceneEnumData" -> "EnumData"

NEstelami commented on ZAPD/GameConfig.cpp:
```diff
@@ -165,6 +165,80 @@ void GameConfig::ConfigFunc_ExternalFile(const tinyxml2::XMLElement& element)
+				cutsceneData.spawnFlag[itemIndex] = itemID;
+
+			if (enumKey == "endSfx")
+				cutsceneData.endSfx[itemIndex] = itemID;
```
Please do `else if` statements here to avoid unnecessary checks.

Yanis002 committed:
add ``else if``s inside ``ConfigFunc_EnumData``

louist103 commented on ZAPD/OtherStructs/CutsceneMM_Commands.cpp:
```diff
@@ -1,53 +1,97 @@
 		entryFmt += element->second.args;
 	}
 
+	if (commandId == CutsceneMM_CommandType::CS_CMD_MISC &&
```
`else if` here too.

louist103 commented on ZAPD/OtherStructs/Cutscene_Common.cpp:
```diff
@@ -0,0 +1,128 @@
+
+CutsceneCommand::~CutsceneCommand()
+{
+	for (auto& entry : entries)
```
This is a vector or pointers so you don't need the reference.

Yanis002 replied:
I'm not sure to understand this one, this is from the file currently called ``Cutscene_Commands.cpp``, do you meant something like this: ``for (auto entry : entries)``?

louist103 replied:
Yes that.

louist103 reviewed (changes requested):
Just these two then good for me.

Yanis002 committed:
more ``else if``s

Yanis002 committed:
removed reference in ``~CutsceneCommand``

louist103 reviewed (approved):
Approved from a ZAPD code perspective. An OOT and MM person may want to verify the docs.
